tests/virt_autotest/kubevirt_tests_server.pm

```diff
@@ -0,0 +1,372 @@
+            my $pre_rel_registry = get_required_var('PREVIOUS_RELEASE_REGISTRY');
+            my $pre_rel_tag = get_required_var('PREVIOUS_RELEASE_TAG');
+            $extra_opt =~ s/PREVIOUS_RELEASE_REGISTRY/$pre_rel_registry/;
+            $extra_opt =~ s/PREVIOUS_RELEASE_TAG/$pre_rel_tag/;
+        }
+
+        $artifacts = "$result_dir/$section";
+        $junit_xml = "$result_dir/${section}.xml";
+        $test_log  = "$result_dir/${section}.log";
+        $test_cmd  = "virt-tests -ginkgo.regexScansFilePath=true ".
+                     "-ginkgo.focus='$go_test' ".
+                     "-ginkgo.skip='QUARANTINE$skip_test' ".
+                     "-ginkgo.slowSpecThreshold 60 ".
+                     "-kubeconfig=/root/.kube/config ".
+                     "-kubectl-path=`which kubectl` ".
+                     "-virtctl-path=`which virtctl` ".
+                     "-installed-namespace=kubevirt ".
+                     "-deploy-testing-infra=false ".
+                     "-config=$test_suite_config ".
+                     "-dns-service-name=rke2-coredns-rke2-coredns ".
+                     "-ginkgo.v=true ".
+                     "-test.v=true ".
+                     "-ginkgo.trace=true ".
+                     "-ginkgo.noisySkippings=false ".
+                     "-ginkgo.progress=true ".
+                     "-ginkgo.noColor ".
+                     "-apply-default-e2e-configuration ".
+                     "$extra_opt ".
+                     "-artifacts=$artifacts ".
+                     "-junit-output=$junit_xml ".
+                     "2>&1 | tee $test_log";
+
+        # my $times;
+        # while ($times && script_run($test_cmd, timeout => 5400)) {
+        #     $times--;
+        # }
+        $retry_times = 1 if ($retry_times);
+        script_retry($test_cmd, retry => $retry_times, delay => 15, timeout => 7200);
+        save_screenshot;
+        send_key 'ctrl-c';
+        assert_script_run("sed -i 's/Tests Suite/$section/g' $junit_xml");
+    }
+}
+
+sub generate_test_report {
+    my $build_xml  = "/tmp/buildTestReports.xml";
+    my $result_dir = "/tmp/artifacts";
+    my $html_dir   = "$result_dir/html";
+
+    script_run("cat > $build_xml <<__END
+<project name=\"genTestReport\" default=\"gen\" basedir=\"$result_dir\">
+    <description>
+        Generate the HTML report from JUnit XML files
+    </description>
+    <target name=\"gen\">
+        <property name=\"genReportDir\" location=\"$result_dir\"/>
+        <delete dir=\"$html_dir\"/>
+        <mkdir dir=\"$html_dir\"/>
+        <junitreport todir=\"$result_dir\">
+            <fileset dir=\"$result_dir\">
+                <include name=\"*_test.xml\" />
+            </fileset>
+            <report format=\"frames\" todir=\"$html_dir\" />
+        </junitreport>
+    </target>
+</project>
+__END
+true");
+
+    # Generate JUnit HTML aggregate test reports
+    script_run("ant -buildfile $build_xml");
+}
+
+sub upload_results {
```

Comment:
It seems your function declaration is not very consistent either. Some have  my ($self, xxx) = @_, some do not.

Reply:
Yes, because some functions did not call any subroutines created in the same test module.